blocks/log/sqlite.py

```diff
@@ -0,0 +1,240 @@
+                                   uuid BLOB NOT NULL,
+                                   "key" text NOT NULL,
+                                   value,
+                                   PRIMARY KEY(uuid, "key")
+                                 );""")
+        self.status = SQLiteStatus(self)
+        super(SQLiteLog, self).__init__(**kwargs)
+
+    def __getstate__(self):
+        """Retrieve the state for pickling.
+
+        :class:`sqlite3.Connection` objects are not picklable, so the
+        `conn` attribute is removed and the connection re-opened upon
+        unpickling.
+
+        """
+        state = self.__dict__.copy()
+        del state['conn']
+        self.resume()
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self.conn = sqlite3.connect(self.database)
+
+    def __getitem__(self, time):
+        self._check_time(time)
+        return SQLiteEntry(self, time)
+
+    def __iter__(self):
+        return map(itemgetter(0), self.conn.execute(
+            ANCESTORS_QUERY + "SELECT DISTINCT time FROM entries "
+            "WHERE uuid IN ancestors ORDER BY time ASC", (self.b_uuid,)
+        ))
+
+    def __len__(self):
+        return self.conn.execute(
+            ANCESTORS_QUERY + "SELECT COUNT(DISTINCT time) FROM entries "
+            "WHERE uuid IN ancestors ORDER BY time ASC", (self.b_uuid,)
+        ).fetchone()[0]
+
+
+class SQLiteStatus(MutableMapping):
+    def __init__(self, log):
+        self.log = log
+
+    def __getitem__(self, key):
+        row = self.log.conn.execute(
+            "SELECT value FROM status WHERE uuid = ? AND key = ?",
+            (self.log.b_uuid, key)
+        ).fetchone()
+        return _get_row(row, key)
+
+    def __setitem__(self, key, value):
+        _register_adapter(value, key)
+        with self.log.conn:
+            self.log.conn.execute(
+                "INSERT OR REPLACE INTO status VALUES (?, ?, ?)",
+                (self.log.b_uuid, key, value)
+            )
+
+    def __delitem__(self, key):
+        with self.log.conn:
+            self.log.conn.execute(
+                "DELETE FROM status WHERE uuid = ? AND key = ?",
+                (self.log.b_uuid, key)
+            )
+
+    def __len__(self):
+        return self.log.conn.execute(
+            "SELECT COUNT(*) FROM status WHERE uuid = ?",
+            (self.log.b_uuid,)
+        ).fetchone()[0]
+
+    def __iter__(self):
+        return map(itemgetter(0), self.log.conn.execute(
+            "SELECT key FROM status WHERE uuid = ?", (self.log.b_uuid,)
+        ))
+
+
+class SQLiteEntry(MutableMapping):
+    """Store log entries in an SQLite database.
+
+    Each entry is a row with the columns `uuid`, `time` (iterations done),
+    `key` and `value`. Note that SQLite only supports numeric values,
+    strings, and bytes (e.g. the `uuid` column). This means there is no
+    support for tuples, dictionaries, NumPy arrays, etc.
```

Comment:
This makes it sounds like you can't use them, whereas... it seems that you can, with all the adapter jazz.